src/test/ui/rfc-2008-non-exhaustive/reachable-patterns.rs

```diff
@@ -0,0 +1,139 @@
+        _ => {}
+    }
+    //~^^ reachable patterns not covered of non exhaustive enum
+
+    // Ok: all covered and not `unreachable-patterns`
+    #[deny(unreachable_patterns)]
+    match non_enum {
+        NonExhaustiveEnum::Unit => {}
+        NonExhaustiveEnum::Tuple(_) => {}
+        NonExhaustiveEnum::Struct { .. } => {}
+        #[deny(non_exhaustive_reachable_patterns)]
+        _ => {}
+    }
+
+    #[deny(non_exhaustive_reachable_patterns)]
+    match NestedNonExhaustive::B {
+        NestedNonExhaustive::A(NonExhaustiveEnum::Unit) => {}
+        NestedNonExhaustive::A(_) => {}
+        NestedNonExhaustive::B => {}
+        _ => {}
+    }
+    //~^^ reachable patterns not covered of non exhaustive enum
+    //~^^^^^ reachable patterns not covered of non exhaustive enum
+
+    // The io::ErrorKind has many `unstable` fields how do they interact with this
+    // lint
+    #[deny(non_exhaustive_reachable_patterns)]
+    match std::io::ErrorKind::Other {
+        std::io::ErrorKind::NotFound => {}
+        std::io::ErrorKind::PermissionDenied => {}
+        std::io::ErrorKind::ConnectionRefused => {}
+        std::io::ErrorKind::ConnectionReset => {}
+        std::io::ErrorKind::ConnectionAborted => {}
+        std::io::ErrorKind::NotConnected => {}
+        std::io::ErrorKind::AddrInUse => {}
+        std::io::ErrorKind::AddrNotAvailable => {}
+        std::io::ErrorKind::BrokenPipe => {}
+        std::io::ErrorKind::AlreadyExists => {}
+        std::io::ErrorKind::WouldBlock => {}
+        std::io::ErrorKind::InvalidInput => {}
+        std::io::ErrorKind::InvalidData => {}
+        std::io::ErrorKind::TimedOut => {}
+        std::io::ErrorKind::WriteZero => {}
+        std::io::ErrorKind::Interrupted => {}
+        std::io::ErrorKind::Other => {}
+        std::io::ErrorKind::UnexpectedEof => {}
+        std::io::ErrorKind::Unsupported => {}
+        std::io::ErrorKind::OutOfMemory => {}
+        // All stable variants are above and unstable in `_`
+        _ => {}
+    }
+    //~^^ reachable patterns not covered of non exhaustive enum
+
+    #[warn(non_exhaustive_reachable_patterns)]
+    match VariantNonExhaustive::Baz(1, 2) {
+        VariantNonExhaustive::Baz(_, _) => {}
+        VariantNonExhaustive::Bar { x, .. } => {}
+    }
+    //~^^ reachable patterns not covered of non exhaustive struct
+
+    #[warn(non_exhaustive_reachable_patterns)]
+    let FunctionalRecord { first_field, second_field, .. } = FunctionalRecord::default();
+    //~^ reachable patterns not covered of non exhaustive struct
+
+    // Ok: this is local
+    #[warn(non_exhaustive_reachable_patterns)]
+    let Foo { a, b, .. } = Foo::default();
+
+    #[warn(non_exhaustive_reachable_patterns)]
+    let NestedStruct { bar: NormalStruct { first_field, .. }, .. } = NestedStruct::default();
+    //~^ reachable patterns not covered of non exhaustive struct
+    //~^^ reachable patterns not covered of non exhaustive struct
+}
```

Comment:
Could you add an `if let` test on an enum to remember that people might be surprised when the lint triggers there? E.g. we want an error on the following:
```rust
#[deny(non_exhaustive_reachable_patterns)]
if let NonExhaustiveEnum::Tuple(_) = non_enum {
}
```

Reply:
UhOh we get a weird error message but it does trigger
```
error: some variants are not matched explicitly
   |
note: the lint level is defined here
  --> /home/devinr/aprog/rust/__forks__/rust/src/test/ui/rfc-2008-non-exhaustive/reachable-patterns.rs:156:12
   |
LL |     #[deny(non_exhaustive_reachable_patterns)]
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: ensure that all variants are matched explicitly by adding the suggested match arms
   = note: the matched value is of type `NonExhaustiveEnum` and the `non_exhaustive_reachable_patterns` attribute was found
```

Reply:
oh that's not great. Is it the HirId or is it the span of the wildcard? Either way we need to find a way to point to the relevant line but I don't know where that info would be